### PR TITLE
Improve OnIdleTasks management

### PR DESCRIPTION
**Pull request by romange**
Signed-off-by: Roman Gershman <romange@gmail.com>

### CODE REVIEW

File: util/proactor_base.cc [hunk 1]
```diff
@@ -66,6 +66,8 @@ ProactorBase::ProactorBase() : task_queue_(512) {
 }
 
 ProactorBase::~ProactorBase() {
+  on_idle_arr_.clear();
+
```
Comment (dranikpg) [inline util/proactor_base.cc:70]: Doesn't this happen on destruction either way 😅

File: util/proactor_base.cc [hunk 2]
```diff
@@ -82,53 +84,63 @@ void ProactorBase::Stop() {
...
 
-void ProactorBase::CancelIdleTask(uint32_t id) {
-  auto it = on_idle_map_.find(id);
-  if (it != on_idle_map_.end()) {
-    on_idle_map_.erase(it);
-    idle_it_ = on_idle_map_.begin();
-  }
+bool ProactorBase::RemoveOnIdleTask(uint32_t id) {
+  if (id < on_idle_arr_.size() || !on_idle_arr_[id].task)
+    return false;
```
Comment (dranikpg) [inline util/proactor_base.cc:139]: The `if  id < on_idle_arr_.size() then return` seems odd. I guess it should be the opposite

File: util/proactor_base.cc [hunk 3]
```diff
@@ -82,53 +84,63 @@ void ProactorBase::Stop() {
...
-    if (!res) {
-      on_idle_map_.erase(idle_it_);
-      idle_it_ = on_idle_map_.begin();
-      break;
+    if (on_idle.task && on_idle.next_ts <= curr_ts) {
+      tl_info_.monotonic_time = curr_ts;
+
+      uint32_t level = on_idle.task();  // run the task
+
+      curr_ts = GetClockNanos();
```
Comment (dranikpg) [inline util/proactor_base.cc:116]: If `GetClockNanos` is non monotonic, then you might end up jumping back in time and even running the same task twice (if it has a high level and a small delta). Or is that highly unlikely? So a `curr_ts = max(curr_ts, GetClockNanos())` would solve it but probably bring other problems. Maybe its even not important at all

File: util/proactor_base.cc (hunk 1 above)
Comment (romange) [reply util/proactor_base.cc:70]: yeah, I guess you are right.

File: util/proactor_base.cc (hunk 3 above)
Comment (romange) [reply util/proactor_base.cc:116]: yeah, this is not what worries me. as long as we won't have weird integer overflow i do not care.
the corner case is when someone changes the system clock. we will need to test it some day anyway.

File: util/proactor_base.cc (hunk 2 above)
Comment (romange) [reply util/proactor_base.cc]: good catch

**Comment by romange** [comment]
@dranikpg ptal

**Review by dranikpg — APPROVED** [review]
